pkg/sql/mutations/mutations.go

```diff
@@ -46,9 +50,131 @@ func ForAllStatements(
 	return sb.String(), changed
 }
 
+// ForeignKeyMutator adds ALTER TABLE ADD FOREIGN KEY statements.
+func ForeignKeyMutator(rng *rand.Rand, stmts []tree.Statement) (additional []tree.Statement) {
+	// Find columns in the tables.
+	cols := map[tree.TableName][]*tree.ColumnTableDef{}
+	byName := map[tree.TableName]*tree.CreateTable{}
+	var tables []*tree.CreateTable
+	for _, stmt := range stmts {
+		table, ok := stmt.(*tree.CreateTable)
+		if !ok {
+			continue
+		}
+		tables = append(tables, table)
+		byName[table.Table] = table
+		for _, def := range table.Defs {
+			switch def := def.(type) {
+			case *tree.ColumnTableDef:
+				cols[table.Table] = append(cols[table.Table], def)
+			}
+		}
+	}
+
+	toNames := func(cols []*tree.ColumnTableDef) tree.NameList {
+		names := make(tree.NameList, len(cols))
+		for i, c := range cols {
+			names[i] = c.Name
+		}
+		return names
+	}
+
+	// We cannot mutate the table definitions themselves because 1) we
+	// don't know the order of dependencies (i.e., table 1 could reference
+	// table 4 which doesn't exist yet) and relatedly 2) we don't prevent
+	// circular dependencies. Instead, add new ALTER TABLE commands to the
+	// end of a list of statements.
+
+	// Create some FKs.
+	for rng.Intn(2) == 0 {
+		// Choose a random table.
+		table := tables[rng.Intn(len(tables))]
+		// Choose a random column subset.
+		fkCols := append([]*tree.ColumnTableDef(nil), cols[table.Table]...)
+		rng.Shuffle(len(fkCols), func(i, j int) {
+			fkCols[i], fkCols[j] = fkCols[j], fkCols[i]
+		})
+		// Pick some randomly short prefix. I'm sure there's a closed
+		// form solution to this with a single call to rng.Intn but I'm
+		// not sure what to search for.
+		i := 1
+		for len(fkCols) > i && rng.Intn(2) == 0 {
+			i++
+		}
+		fkCols = fkCols[:i]
+
+		// Check if a table has the needed column types.
+	LoopTable:
+		for refTable, refCols := range cols {
```

Comment:
I found an error that showed up in my ci (I think i sent it to you already), but i don't think you can use the same column in multiple foreign key constraints? This threw an error in some randomly generated create table statement.

Reply:
I added a commit to address this. PTAL.